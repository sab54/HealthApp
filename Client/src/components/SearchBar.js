// Client/src/components/SearchBar.js
/**
 * SearchBar.js
 * 
 * This file defines the `SearchBar` component, which provides a customizable search 
 * input with features such as debounce handling, voice input (optional), and a clear button. 
 * The search input is designed with animated effects to enhance the user experience when 
 * focusing, typing, or clearing the input. The component can be themed based on the `theme` 
 * prop, and it supports accessibility for better usability.
 * 
 * Features:
 * - Debounced input change to limit unnecessary calls during typing (via lodash `debounce`).
 * - Clear button that appears when there is input and can reset the search field.
 * - Voice input button (optional) to trigger voice search (functionality placeholder).
 * - Animated border and button opacity that responds to focus and input changes.
 * - Accessibility features, including labels and hints for better screen reader support.
 * 
 * Props:
 * - `query`: The initial search query value (used to set the default value of the input).
 * - `onChange`: Callback function to handle input changes.
 * - `theme`: The theme object to customize colors for background, text, borders, etc.
 * - `placeholder`: Placeholder text for the input field (default is "Search...").
 * - `debounceTime`: Time (in milliseconds) to debounce input changes (default is 300ms).
 * - `onSubmit`: Function triggered when the search is submitted (via "Enter" or "Search" button).
 * - `showVoice`: Boolean to control whether the voice input button is displayed.
 * 
 * Dependencies:
 * - `react-native`
 * - `@expo/vector-icons`
 * - `lodash`
 * 
 * Author: Sunidhi Abhange
 */

import React, { useState, useEffect, useRef } from 'react';
import {
    View,
    TextInput,
    StyleSheet,
    TouchableOpacity,
    Animated,
    Keyboard,
} from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import { debounce } from 'lodash';

const SearchBar = ({
    query,
    onChange,
    theme,
    placeholder = 'Search...',
    debounceTime = 300,
    onSubmit = () => {},
    showVoice = false,
}) => {
    const [isFocused, setIsFocused] = useState(false);
    const [inputValue, setInputValue] = useState(query);
    const inputRef = useRef(null);
    const animatedBorder = useRef(new Animated.Value(0)).current;
    const clearOpacity = useRef(new Animated.Value(0)).current;

    const debouncedChange = useRef(
        debounce((text) => onChange(text), debounceTime)
    ).current;

    useEffect(() => {
        debouncedChange(inputValue);
    }, [inputValue]);

    useEffect(() => {
        Animated.timing(animatedBorder, {
            toValue: isFocused ? 1 : 0,
            duration: 200,
            useNativeDriver: false,
        }).start();
    }, [isFocused]);

    useEffect(() => {
        Animated.timing(clearOpacity, {
            toValue: inputValue.length > 0 ? 1 : 0,
            duration: 200,
            useNativeDriver: true,
        }).start();
    }, [inputValue]);

    const borderColor = animatedBorder.interpolate({
        inputRange: [0, 1],
        outputRange: [
            theme.card || '#eee',
            theme.accent || theme.primary || '#4B7BE5',
        ],
    });

    const handleClear = () => {
        setInputValue('');
        onChange('');
        inputRef.current?.focus();
    };

    const handleVoiceInput = () => {
        console.log('Voice input placeholder logic');
    };

    return (
        <Animated.View
            style={[
                styles.wrapper,
                {
                    backgroundColor: theme.input || theme.card || '#fff',
                    borderColor,
                },
            ]}
        >
            <Ionicons name='search' size={20} color={theme.text || '#000'} />

            <TextInput
                ref={inputRef}
                style={[styles.input, { color: theme.text || '#000' }]}
                placeholder={placeholder}
                placeholderTextColor={(theme.text || '#000') + '99'}
                value={inputValue}
                onChangeText={setInputValue}
                onFocus={() => setIsFocused(true)}
                onBlur={() => setIsFocused(false)}
                returnKeyType='search'
                onSubmitEditing={() => {
                    onSubmit(inputValue);
                    Keyboard.dismiss();
                }}
                accessibilityLabel='Search input'
                accessibilityHint='Enter text to search'
                accessibilityRole='search'
            />

            <Animated.View style={{ opacity: clearOpacity }}>
                {inputValue.length > 0 && (
                    <TouchableOpacity onPress={handleClear}>
                        <Ionicons
                            name='close-circle'
                            size={20}
                            color={(theme.text || '#000') + '88'}
                        />
                    </TouchableOpacity>
                )}
            </Animated.View>

            {showVoice && (
                <TouchableOpacity
                    onPress={handleVoiceInput}
                    accessibilityLabel='Voice input'
                    accessibilityHint='Activate voice search'
                    style={{ marginLeft: 8 }}
                >
                    <Ionicons
                        name='mic-outline'
                        size={20}
                        color={(theme.text || '#000') + '88'}
                    />
                </TouchableOpacity>
            )}
        </Animated.View>
    );
};

const styles = StyleSheet.create({
    wrapper: {
        flexDirection: 'row',
        alignItems: 'center',
        borderRadius: 10,
        paddingVertical: 4,
        paddingHorizontal: 8,
        marginBottom: 10,
        borderWidth: 2,
    },
    input: {
        flex: 1,
        fontSize: 13,
        marginLeft: 8,
        fontFamily: 'Poppins',
        paddingVertical: 2,
    },
});

export default SearchBar;
