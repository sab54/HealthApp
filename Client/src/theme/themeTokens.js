export const getThemeColors = (isDarkMode) => ({
    mode: isDarkMode ? 'dark' : 'light',

    // Layout & Backgrounds
    background: isDarkMode ? '#1E1B2E' : '#ffffff', // App root background
    card: isDarkMode ? '#2A2340' : '#f9f9ff', // Cards or blocks background
    surface: isDarkMode ? '#26203A' : '#f5f5ff', // Neutral surfaces (modals, panels)
    highlight: isDarkMode ? '#3A2D5C' : '#e2e2ff', // Hoverable elements
    overlay: isDarkMode ? 'rgba(0,0,0,0.7)' : 'rgba(255,255,255,0.6)', // Modal overlays

    // Text & Typography
    title: isDarkMode ? '#ffffff' : '#1a1a2e', // Main headings
    text: isDarkMode ? '#D1D1FF' : '#333366', // Standard body text
    mutedText: isDarkMode ? '#A3A3C2' : '#7d7da6', // Less prominent text
    placeholder: isDarkMode ? 'rgba(255,255,255,0.4)' : 'rgba(0,0,128,0.3)', // Input placeholder

    // Inputs & Borders
    input: isDarkMode ? '#2A2340' : '#ffffff', // Text input background
    inputText: isDarkMode ? '#E5E5FF' : '#333366', // Input text color
    inputBorder: isDarkMode ? '#5C4B8A' : '#c5c5ff', // Input border
    border: isDarkMode ? '#5C4B8A' : '#c5c5ff', // General borders
    focusOutline: isDarkMode ? '#7C5DFF' : '#6C5BCD', // Focus ring color

     // Checkbox styles
    checkboxBorder: isDarkMode ? '#6C5BCD' : '#9999cc', // Subtle border
    checkboxBackground: isDarkMode ? '#3A2D5C' : '#ffffff',  // Dark background for unchecked state
    checkboxChecked: '#6C5BCD', // Purple when checked
    checkboxHover: isDarkMode ? '#3A2D5C' : '#e2e2ff',  // Hover effect
    checkboxTick: isDarkMode ? '#ffffffff' : '#ffffffff',  // White for light mode, purple for dark mode

    // Buttons
    buttonPrimaryBackground: '#6C5BCD', // Purple primary button
    buttonPrimaryText: '#ffffff',
    buttonSecondaryBackground: '#4B9BE3', // Blue secondary button
    buttonSecondaryText: '#ffffff',
    buttonDisabledBackground: isDarkMode ? '#3A324D' : '#cccccc',
    buttonDisabledText: isDarkMode ? '#7D7DA6' : '#999999',

    // radio & Checkbox
    radioBorder: isDarkMode ? '#7D7DA6' : '#9999cc',
    radioBackground: isDarkMode ? '#2A2340' : '#ffffff',
    radioChecked: '#6C5BCD',
    radioHover: isDarkMode ? '#3A2D5C' : '#e2e2ff',
    radioDisabled: isDarkMode ? '#5C4B8A' : '#cccccc',

    // Status Feedback Colors
    success: '#4CAF50',
    successBackground: isDarkMode ? '#234B2D' : '#d4edda',
    error: '#D93F2B',
    errorBackground: isDarkMode ? '#4B2D2B' : '#f8d7da',
    warning: '#FFB900',
    warningBackground: isDarkMode ? '#4F4220' : '#fff3cd',
    info: '#4B9BE3', // Blue info
    infoBackground: isDarkMode ? '#243B52' : '#d1ecf1',

    // Actions & Interactions
    actionBackground: isDarkMode ? '#26203A' : '#f4f4ff',
    actionText: isDarkMode ? '#D1D1FF' : '#333366',
    disabled: isDarkMode ? '#3A324D' : '#e0e0e0',

    // Icons & Visuals
    icon: isDarkMode ? '#C5C5FF' : '#4a4a8a',

    // Links & Interactivity
    link: '#4B9BE3',
    linkHover: '#3A7ACC',

    // Badges & Tags
    badge: '#6C5BCD',
    badgeText: '#ffffff',
    tagBackground: isDarkMode ? '#3A2D5C' : '#ebf2ff',
    tagText: isDarkMode ? '#E0E0FF' : '#1a1a2e',

    // Dividers & Shadows
    divider: isDarkMode ? '#5C4B8A' : '#c5c5ff',
    shadow: isDarkMode ? '#00000066' : '#0000001a',
    cardShadow: isDarkMode ? '#00000066' : '#0000001a',

    // Modals & Popups
    modalBackground: isDarkMode ? '#2A2340' : '#ffffff',

    // Hover/Focus Effects
    hoverBackground: isDarkMode ? '#3A2D5C' : '#f4f4ff',

    // Header & Footer Styling
    headerBackground: isDarkMode ? '#26203A' : '#f9f9ff',
    footerBackground: isDarkMode ? '#1E1B2E' : '#ffffff',
});
