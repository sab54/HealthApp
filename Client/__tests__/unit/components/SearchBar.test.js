/**
 * SearchBar.test.js
 *
 * What This Test File Covers:
 *
 * 1. Basic Rendering
 *    - Renders input, icons, and placeholder.
 *
 * 2. Input Behavior
 *    - Handles input changes and triggers debounced onChange.
 *    - Calls onSubmitEditing on return.
 *
 * 3. Clear Button
 *    - Appears when input is non-empty.
 *    - Clears and refocuses input.
 *
 * 4. Voice Input Icon
 *    - Renders mic icon when `showVoice` is true.
 *
 * 5. UI & Theming
 *    - Applies light/dark theme correctly.
 *
 * 6. Accessibility
 *    - Text input is accessible.
 *
 * 7. Snapshots
 *    - Matches snapshot for light and dark theme.
 *
 * 8. Animation Styles
 *    - Clear icon animates in with opacity.
 */

import React from 'react';
import { render, fireEvent, act } from '@testing-library/react-native';
import SearchBar from '@/components/SearchBar';
import { getThemeColors } from '@/theme/themeTokens';

jest.useFakeTimers();

// Mock Ionicons
jest.mock('@expo/vector-icons', () => {
  const React = require('react');
  const { Text } = require('react-native');
  return {
    Ionicons: ({ name }) => <Text>{`Icon: ${name}`}</Text>,
  };
});

const lightTheme = getThemeColors(false);
const darkTheme = getThemeColors(true);

describe('SearchBar Component', () => {
  const onChangeMock = jest.fn();
  const onSubmitMock = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with placeholder and icons', () => {
    const { getByPlaceholderText, getByText } = render(
      <SearchBar
        query=""
        onChange={onChangeMock}
        onSubmit={onSubmitMock}
        theme={lightTheme}
      />
    );

    expect(getByPlaceholderText('Search...')).toBeTruthy();
    expect(getByText('Icon: search')).toBeTruthy();
  });

  it('updates input and debounces onChange', () => {
    const { getByPlaceholderText } = render(
      <SearchBar
        query=""
        onChange={onChangeMock}
        onSubmit={onSubmitMock}
        theme={lightTheme}
      />
    );

    const input = getByPlaceholderText('Search...');
    fireEvent.changeText(input, 'query');

    act(() => {
      jest.advanceTimersByTime(300);
    });

    expect(onChangeMock).toHaveBeenCalledWith('query');
  });

  it('submits on return key press', () => {
    const { getByPlaceholderText } = render(
      <SearchBar
        query=""
        onChange={onChangeMock}
        onSubmit={onSubmitMock}
        theme={lightTheme}
      />
    );

    const input = getByPlaceholderText('Search...');
    fireEvent.changeText(input, 'run');
    fireEvent(input, 'submitEditing');

    expect(onSubmitMock).toHaveBeenCalledWith('run');
  });

  it('clears input and refocuses when clear button pressed', () => {
    const { getByPlaceholderText, getByText } = render(
      <SearchBar
        query="clearme"
        onChange={onChangeMock}
        onSubmit={onSubmitMock}
        theme={lightTheme}
      />
    );

    const input = getByPlaceholderText('Search...');
    fireEvent.changeText(input, 'clearme');

    act(() => {
      jest.advanceTimersByTime(300);
    });

    const clearIcon = getByText('Icon: close-circle');
    fireEvent.press(clearIcon);

    expect(onChangeMock).toHaveBeenCalledWith('');
  });

  it('renders mic icon if showVoice is true', () => {
    const { getByText } = render(
      <SearchBar
        query=""
        onChange={onChangeMock}
        onSubmit={onSubmitMock}
        showVoice
        theme={lightTheme}
      />
    );

    expect(getByText('Icon: mic-outline')).toBeTruthy();
  });

  it('applies dark theme styling', () => {
    const { getByPlaceholderText } = render(
      <SearchBar
        query="dark"
        onChange={onChangeMock}
        onSubmit={onSubmitMock}
        theme={darkTheme}
      />
    );

    const input = getByPlaceholderText('Search...');
    expect(input.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ color: darkTheme.text }),
      ])
    );
  });

  it('is accessible to screen readers', () => {
    const { getByPlaceholderText } = render(
      <SearchBar
        query=""
        onChange={onChangeMock}
        onSubmit={onSubmitMock}
        theme={lightTheme}
      />
    );

    const input = getByPlaceholderText('Search...');
    expect(input.props.accessibilityRole).toBe('search');
    expect(input.props.accessibilityLabel).toBe('Search input');
  });

  it('matches snapshot (light theme)', () => {
    const { toJSON } = render(
      <SearchBar
        query="light"
        onChange={onChangeMock}
        onSubmit={onSubmitMock}
        theme={lightTheme}
      />
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot (dark theme)', () => {
    const { toJSON } = render(
      <SearchBar
        query="dark"
        onChange={onChangeMock}
        onSubmit={onSubmitMock}
        theme={darkTheme}
      />
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('animates clear icon opacity', () => {
    const { toJSON, getByPlaceholderText } = render(
      <SearchBar
        query="test"
        onChange={onChangeMock}
        onSubmit={onSubmitMock}
        theme={lightTheme}
      />
    );

    const input = getByPlaceholderText('Search...');
    fireEvent.changeText(input, '123');

    act(() => {
      jest.advanceTimersByTime(200);
    });

    const tree = toJSON();
    const animatedView = tree.children.find(
      (child) => child?.type === 'View' && child?.props?.style?.opacity !== undefined
    );

    expect(animatedView?.props?.style?.opacity).toBeDefined();
  });
});
