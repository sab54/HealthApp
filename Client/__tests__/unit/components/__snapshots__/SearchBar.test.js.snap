// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Component matches snapshot (dark theme) 1`] = `
<View
  collapsable={false}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#2A2340",
      "borderColor": "rgba(42, 35, 64, 1)",
      "borderRadius": 10,
      "borderWidth": 2,
      "flexDirection": "row",
      "marginBottom": 10,
      "paddingHorizontal": 8,
      "paddingVertical": 4,
    }
  }
>
  <Text>
    Icon: search
  </Text>
  <TextInput
    accessibilityHint="Enter text to search"
    accessibilityLabel="Search input"
    accessibilityRole="search"
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Search..."
    placeholderTextColor="#D1D1FF99"
    returnKeyType="search"
    style={
      [
        {
          "flex": 1,
          "fontFamily": "Poppins",
          "fontSize": 13,
          "marginLeft": 8,
          "paddingVertical": 2,
        },
        {
          "color": "#D1D1FF",
        },
      ]
    }
    value="dark"
  />
  <View
    collapsable={false}
    style={
      {
        "opacity": 0,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text>
        Icon: close-circle
      </Text>
    </View>
  </View>
</View>
`;

exports[`SearchBar Component matches snapshot (light theme) 1`] = `
<View
  collapsable={false}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#ffffff",
      "borderColor": "rgba(249, 249, 255, 1)",
      "borderRadius": 10,
      "borderWidth": 2,
      "flexDirection": "row",
      "marginBottom": 10,
      "paddingHorizontal": 8,
      "paddingVertical": 4,
    }
  }
>
  <Text>
    Icon: search
  </Text>
  <TextInput
    accessibilityHint="Enter text to search"
    accessibilityLabel="Search input"
    accessibilityRole="search"
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Search..."
    placeholderTextColor="#33336699"
    returnKeyType="search"
    style={
      [
        {
          "flex": 1,
          "fontFamily": "Poppins",
          "fontSize": 13,
          "marginLeft": 8,
          "paddingVertical": 2,
        },
        {
          "color": "#333366",
        },
      ]
    }
    value="light"
  />
  <View
    collapsable={false}
    style={
      {
        "opacity": 0,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text>
        Icon: close-circle
      </Text>
    </View>
  </View>
</View>
`;
